src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1381,6 +1381,63 @@
   (let [last-modified-milliseconds (.lastModified (io/file cacrl))]
        (time-coerce/from-long last-modified-milliseconds)))
 
+(schema/defn ^:always-validate reject-delta-crl
+  [crl :- X509CRL]
+  (when (utils/delta-crl? crl)
+    (throw (IllegalArgumentException.
+            ^String (i18n/trs "Cannot support delta CRL.")))))
+
+(schema/defn ^:always-validate validate-certs-and-crls
+  "Given a list of certificates and a list of CRLs, validate the certificate
+   chain, i.e. ensure that none of the certs have been revoked by checking the
+   appropriate CRL, which must be present and currently valid. Delta CRLs are
+   not supported. Returns nil if successful."
+  [cert-chain :- [X509Certificate]
+   crl-chain :- [X509CRL]]
+  (try
+    (utils/validate-cert-chain cert-chain crl-chain)
+    (catch CertPathValidatorException e
+      (throw (IllegalArgumentException.
+              ^String (i18n/trs "Invalid certs and/or CRLs: {0}" (.getMessage e))))))
+  (doseq [crl crl-chain]
+    (reject-delta-crl crl)))
+
+(schema/defn ^:always-validate get-newest-crl :- X509CRL
+  "Determine the newest CRL by looking for the highest CRL Number. This assumes
+  all CRLs have the same issuer. Fails if more than one CRL has the highest CRL
+  Number."
+  [crls :- [X509CRL]]
+  (let [newest-crl (->> crls
+                        (group-by utils/get-crl-number)
+                        (into (sorted-map))
+                        last
+                        last)]
+    (if (= 1 (count newest-crl))
+      (first newest-crl)
+      (throw (IllegalArgumentException.
+              ^String (i18n/trs "Could not determine newest CRL."))))))
+
+(schema/defn ^:always-validate update-crl :- CertificateRevocationList
+  "Given a collection of CRLs, update the CRL chain and confirm that
+  all CRLs are currently valid."
+  [incoming-crl-pem :- InputStream
+   crl-path :- schema/Str
+   cert-chain-path :- schema/Str]
+  (log/info (i18n/trs "Updating CRL at {0}" crl-path))
+  (let [incoming-crls (utils/pem->crls incoming-crl-pem)
+        current-crls (utils/pem->crls crl-path)
+        cert-chain (utils/pem->certs cert-chain-path)
+        crls-by-issuer (->> incoming-crls
+                            (concat current-crls)
+                            set
```

Comment:
What does set do here?

Reply:
Making sure there aren't duplicates. Like if someone submits their whole CRL chain, but not all the individual CRLs have changed. The way I currently have it implemented, if there were duplicates, the update would fail because it would see two CRLs from the same issuer with the same CRL Number.

Reply:
Wow. I would have assumed that pem->certs would create different objects from those different streams. That's neat.

Reply:
Does it not? That's also surprising to me.

Reply:
It might, I didn't test it, but calling `set` seemed to be a good idea just in case? ¯\_(ツ)_/¯

Reply:
It does work like Molly mentioned, I tested it before I replied.